### PR TITLE
feat: Automerge text integration

### PR DESCRIPTION
**Pull request by dmaretskyi**
### Summary
🔄📝🚀

This pull request refactors the code to use a new abstraction layer for text objects, based on the `DocAccessor` type and the `AutomergeObject` class, which enable switching between Automerge and Yjs backends. It modifies several files in the `@dxos/echo-schema`, `@dxos/react-ui-editor`, and `@dxos/react-client` packages to use the new types and classes, and adds or updates some stories and dependencies. It also fixes some type errors and deprecates some unused functions.

> _Sing, O Muse, of the mighty refactor, that changed the text objects' core_
> _And made them work with different backends, Automerge and Yjs, and more_
> _With `DocAccessor` and `AutomergeObject`, the new abstractions they devised_
> _And with imports, types, and hooks, the old code they revised_

### Walkthrough
*  Refactor the code to use the new `DocAccessor` type and the `AutomergeObject` class to support different backends for text objects. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL185-R185),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL191-R191),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL213-R213),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL280-R281),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL308-R308),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L457-R509),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L558-R562),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L565-R582),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86b6148b95be2f77ac6eff05a19c3bf6926da19271cd10a1db4d9148ce39b9d2L272-R272),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-cb02d960130d98c4aeb32386c5603e3964aab969002a3c0df9e17ce1fa802d3eR5),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-cb02d960130d98c4aeb32386c5603e3964aab969002a3c0df9e17ce1fa802d3eL10-R40),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-874cfefd5ce6f29a1bbd0df27acce8a2aa03e20f12db8d444a09c6802006bc92R653-R655),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-874cfefd5ce6f29a1bbd0df27acce8a2aa03e20f12db8d444a09c6802006bc92L669-R681),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-ba0c70c7f8386fa487925bdae1b38cbd4936d08912dd605bd5e3d069609d494aR27-R29),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86507cc1a9e67c9ce58220dd10bd4fa87bde020fa2816402c93e4bc045c12f78L29-R32),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86507cc1a9e67c9ce58220dd10bd4fa87bde020fa2816402c93e4bc045c12f78L68-R70),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-00523e468a83c67e41036dbef700d4b5af41aa31a7df19234115d989dc26b686L87-R87),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b6976ec4f17913eb677b8888bf25c596147239c9fbba4c38792ed27e5576de5cL40-R44),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b6976ec4f17913eb677b8888bf25c596147239c9fbba4c38792ed27e5576de5cR76-R79),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5R10),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5R21),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5L91-R98),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-54cb076dc9fe4efd73ca6b5742b9b49612a6385331897fbbd76f57a5c5f25fcfL34-R34),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139R10),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139R16),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139L22-R32),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL8-R15),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL21-R28),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aR47-R50),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL45-R81))
  * Add an `opts` parameter to the `TextObject` constructor and pass the `useAutomergeBackend` option to determine whether to use the Automerge backend or the Yjs backend for the text object. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86b6148b95be2f77ac6eff05a19c3bf6926da19271cd10a1db4d9148ce39b9d2L272-R272),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-cb02d960130d98c4aeb32386c5603e3964aab969002a3c0df9e17ce1fa802d3eL10-R40))
  * Add a `handle` property to the `DocAccessor` type that encapsulates the methods for accessing and modifying the underlying Automerge document. Replace the `docAccess.docSync()` and `docAccess.change()` method calls with `docAccess.handle.docSync()` and `docAccess.handle.change()` in the `AutomergeStoreAdapter` class. Replace the `docAccess.addListener()` and `docAccess.removeListener()` method calls with `docAccess.handle.addListener()` and `docAccess.handle.removeListener()` in the `AutomergeStoreAdapter` class. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL185-R185),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL191-R191),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL213-R213),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL280-R281),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-9e5c9b00da0a2c064670c6d8c2243be71a0fae80954aefda3a5cd68f675cac6dL308-R308),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L457-R509))
  * Add a new type `DocAccessor` that extends the `IDocHandle` type with a `path` property and an `isAutomergeDocAccessor` property. Rename the `DocAccessor` type to `IDocHandle` in the `automerge-object.ts` file. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L558-R562),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-3515b2ff773cf8bd489c72171f771f97e2f2a47b3c6b2656e485c0c8db3c5b48L565-R582))
  * Add an `isDocAccessor` type to check if the content is a `DocAccessor` object. Use the `isDocAccessor` type in the `TextEditor.tsx`, `useCollaboration.ts`, and `useTextModel.ts` files. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-ba0c70c7f8386fa487925bdae1b38cbd4936d08912dd605bd5e3d069609d494aR27-R29),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5R21),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139R10),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL8-R15))
  * Add an `automergePlugin` function to create an extension for the editor that enables collaboration using the Automerge backend. Use the `automergePlugin` function in the `useCollaboration.ts` file. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139R16),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-8f8c3f3719fc2a65e7863cc8d93f0dbbf0270a72957dbdeb815c4e32b5ab0139L22-R32))
  * Add the `DocAccessor` type to the union type of the `content` property of the `EditorModel` type. Fix a type error caused by the change of the `content` type in the `yjs.stories.tsx` file. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-54cb076dc9fe4efd73ca6b5742b9b49612a6385331897fbbd76f57a5c5f25fcfL34-R34),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL21-R28))
  * Add a conditional block that returns an `AutomergeObject` instance if the `useAutomergeBackend` option is true in the `TextObject` constructor. Use the `AutomergeObject` class to access the underlying Automerge document. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-cb02d960130d98c4aeb32386c5603e3964aab969002a3c0df9e17ce1fa802d3eL10-R40),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL8-R15))
  * Add a conditional block that assigns the `initialText` variable to the text content from the Automerge document if the content is a `DocAccessor` object in the `BaseTextEditor` component. Use the `get` function from the `lodash.get` package to access the text content from the Automerge document. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5R10),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b00930ab8b3935dc62edbd2f9df9136934286d2c57d50d8c45f260a6fb3a84d5L91-R98))
  * Add a conditional block that returns undefined if the text object is an `AutomergeObject` in the `useTextModel` hook. Skip the creation of the `SpaceAwarenessProvider` for the Automerge backend, since it is not supported. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aR47-R50))
  * Add a `content` variable that is assigned to the text content from the Automerge document if the text object is an `AutomergeObject` in the `useTextModel` hook. Return a text model with the `content` variable and no `provider` if the text object is an `AutomergeObject`. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-68ba916c19a77373445d25441edf86cfed3d47fb79112aedd237068be6c4752aL45-R81))
  * Add some import statements for the `@dxos/automerge` and `@dxos/echo-schema` packages in the `package.json` file of the `react-ui-editor` package. Move these packages from the `devDependencies` section to the `dependencies` section. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86507cc1a9e67c9ce58220dd10bd4fa87bde020fa2816402c93e4bc045c12f78L29-R32),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-86507cc1a9e67c9ce58220dd10bd4fa87bde020fa2816402c93e4bc045c12f78L68-R70))
  * Add an `automerge` parameter to the `Story` component and pass the `useAutomergeBackend` option to the `TextObject` constructor in the `TextEditor.stories.tsx` file. Add a new export `Automerge` that renders the `Story` component with the `automerge` parameter set to true. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b6976ec4f17913eb677b8888bf25c596147239c9fbba4c38792ed27e5576de5cL40-R44),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-b6976ec4f17913eb677b8888bf25c596147239c9fbba4c38792ed27e5576de5cR76-R79))
  * Rename the `EditorStory` export to `AutomergeRepo` in the `automerge.stories.tsx` file. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-00523e468a83c67e41036dbef700d4b5af41aa31a7df19234115d989dc26b686L87-R87))
  * Add a `@deprecated` comment to the `isActualTypedObject` and `isActualAutomergeObject` functions in the `typed-object.ts` file. Mark these functions as deprecated, since they are no longer used in the code. ([link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-874cfefd5ce6f29a1bbd0df27acce8a2aa03e20f12db8d444a09c6802006bc92R653-R655),[link](https://github.com/dxos/dxos/pull/4923/files?diff=unified&w=0#diff-874cfefd5ce6f29a1bbd0df27acce8a2aa03e20f12db8d444a09c6802006bc92L669-R681))

### CODE REVIEW

File: packages/ui/react-ui-editor/src/components/TextEditor/yjs.stories.tsx [hunk 1]
```diff
@@ -31,7 +31,7 @@ const Story = () => {
 
   useDataGenerator({
     generator: generate ? textGenerator : undefined,
-    options: { text: typeof model?.content !== 'string' ? model?.content : undefined },
+    options: { text: typeof model?.content !== 'string' ? (model?.content as any) : undefined },
```
Comment (wittjosiah) [inline packages/ui/react-ui-editor/src/components/TextEditor/yjs.stories.tsx:34]: why? temp? comment?

File: packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx [hunk 2]
```diff
@@ -37,11 +37,11 @@ const hyperLinkTooltip = () =>
     );
   });
 
-const Story = () => {
-  const [item] = useState({ text: new TextObject(text) });
+const Story = ({ automerge }: { automerge?: boolean }) => {
```
Comment (wittjosiah) [inline packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx:40]: this is kinda awkward, if there's common stuff I'd factor it out to a separate component

File: packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx [hunk 3]
```diff
@@ -37,11 +37,11 @@ const hyperLinkTooltip = () =>
...
+  const [item] = useState({ text: new TextObject(text, undefined, undefined, { useAutomergeBackend: !!automerge }) });
   const model = useTextModel({ text: item.text });
   if (!model) {
-    return null;
+    return <></>;
```
Comment (wittjosiah) [inline packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx:44]: why not null?

File: packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx (hunk 2 above)
Comment (dmaretskyi) [reply packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx:40]: I've tried using storybook's controlls for this, but got stuck on Story's vs component's prop types missmatch. Can we do a session together where you show me the best prcatices for using storybooks?

File: packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx (hunk 3 above)
Comment (dmaretskyi) [reply packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.stories.tsx:44]: Storybook types didn't like null return type for component

**Pull request merged**